Dub method for dependencies

The dependency search for Dub method dependencies always seems to find that the library was built with dmd even when it was built with ldc2.
```
Found DUB: /usr/bin/dub (DUB version 1.10.0, built on Jul  3 2018)
ERROR: Dependency unit-threaded found but it was compiled with dmd and we are using ldc
Dependency unit-threaded found: NO 

meson.build:38:0: ERROR:  Dependency "unit-threaded" not found, tried 
```

Update: building everything with dmd does work, building with ldc2 always fails.

It is also worth noting that Meson only uses LDC options for a build. I build the dependency using Dub and DMD, and then set up the project to use DMD (DC must be specified otherwise /usr/bin/ldc2 is used)
```
|> DC=/usr/bin/dmd meson_head --prefix=$HOME/Built ~/Repositories/Git/Masters/GFontBrowser_D
The Meson build system
Version: 0.48.0.dev1
Source dir: /home/users/russel/Repositories/Git/Masters/GFontBrowser_D
Build dir: /home/users/russel/BuildArea/GFontBrowser_D
Build type: native build
Project name: GFontBrowser
Project version: 0.0.0
Native D compiler: /usr/bin/dmd (dmd 2.081.1 "DMD64 D Compiler v2.081.1")
Build machine cpu family: x86_64
Build machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (0.29)
Dependency gtkd-3 found: YES 3.8.3
Found DUB: /usr/bin/dub (DUB version 1.10.0, built on Jul  3 2018)
Dependency unit-threaded found: YES 0.7.50
Build targets in project: 2
Found ninja-1.8.2 at /usr/bin/ninja
```
but then on building:
```
|> ninja
[0/1] Regenerating build files.
The Meson build system
Version: 0.48.0.dev1
Source dir: /home/users/russel/Repositories/Git/Masters/GFontBrowser_D
Build dir: /home/users/russel/BuildArea/GFontBrowser_D
Build type: native build
Project name: GFontBrowser
Project version: 0.0.0
Native D compiler: /usr/bin/dmd (dmd 2.081.1 "DMD64 D Compiler v2.081.1")
Build machine cpu family: x86_64
Build machine cpu: x86_64
Dependency gtkd-3 found: YES (cached)
Dependency unit-threaded found: YES (cached)
Build targets in project: 2
Found ninja-1.8.2 at /usr/bin/ninja
[1/8] Compiling D object 'gfontbrowser@exe/source_GFontBrowser.d.o'.
FAILED: gfontbrowser@exe/source_GFontBrowser.d.o 
/usr/bin/dmd -Igfontbrowser@exe -I. -I../../Repositories/Git/Masters/GFontBrowser_D -I../../Repositories/Git/Masters/GFontBrowser_D/source -I/usr/include/d/gtkd-3/ -color=on -wi -O -release -O3 -J/home/users/russel/Repositories/Git/Masters/GFontBrowser_D   -of'gfontbrowser@exe/source_GFontBrowser.d.o' -c ../../Repositories/Git/Masters/GFontBrowser_D/source/GFontBrowser.d
Error: unrecognized switch '-O3'
       run 'dmd -man' to open browser on manual
[2/8] Compiling D object 'gfontbrowser@exe/source_main.d.o'.
FAILED: gfontbrowser@exe/source_main.d.o 
/usr/bin/dmd -Igfontbrowser@exe -I. -I../../Repositories/Git/Masters/GFontBrowser_D -I../../Repositories/Git/Masters/GFontBrowser_D/source -I/usr/include/d/gtkd-3/ -color=on -wi -O -release -O3 -J/home/users/russel/Repositories/Git/Masters/GFontBrowser_D   -of'gfontbrowser@exe/source_main.d.o' -c ../../Repositories/Git/Masters/GFontBrowser_D/source/main.d
Error: unrecognized switch '-O3'
       run 'dmd -man' to open browser on manual
[3/8] Compiling D object 'gfontbrowser@exe/source_About.d.o'.
FAILED: gfontbrowser@exe/source_About.d.o 
/usr/bin/dmd -Igfontbrowser@exe -I. -I../../Repositories/Git/Masters/GFontBrowser_D -I../../Repositories/Git/Masters/GFontBrowser_D/source -I/usr/include/d/gtkd-3/ -color=on -wi -O -release -O3 -J/home/users/russel/Repositories/Git/Masters/GFontBrowser_D   -of'gfontbrowser@exe/source_About.d.o' -c ../../Repositories/Git/Masters/GFontBrowser_D/source/About.d
Error: unrecognized switch '-O3'
       run 'dmd -man' to open browser on manual
[4/8] Compiling D object 'gfontbrowser-test@exe/source_About.d.o'.
FAILED: gfontbrowser-test@exe/source_About.d.o 
/usr/bin/dmd -Igfontbrowser-test@exe -I. -I../../Repositories/Git/Masters/GFontBrowser_D -I../../Repositories/Git/Masters/GFontBrowser_D/source -I/usr/include/d/gtkd-3/ -I/home/users/russel/.dub/packages/unit-threaded-0.7.50/unit-threaded/source/ -color=on -wi -O -release -O3 -J/home/users/russel/Repositories/Git/Masters/GFontBrowser_D -unittest   -of'gfontbrowser-test@exe/source_About.d.o' -c ../../Repositories/Git/Masters/GFontBrowser_D/source/About.d
Error: unrecognized switch '-O3'
       run 'dmd -man' to open browser on manual
[5/8] Compiling D object 'gfontbrowser-test@exe/source_main.d.o'.
FAILED: gfontbrowser-test@exe/source_main.d.o 
/usr/bin/dmd -Igfontbrowser-test@exe -I. -I../../Repositories/Git/Masters/GFontBrowser_D -I../../Repositories/Git/Masters/GFontBrowser_D/source -I/usr/include/d/gtkd-3/ -I/home/users/russel/.dub/packages/unit-threaded-0.7.50/unit-threaded/source/ -color=on -wi -O -release -O3 -J/home/users/russel/Repositories/Git/Masters/GFontBrowser_D -unittest   -of'gfontbrowser-test@exe/source_main.d.o' -c ../../Repositories/Git/Masters/GFontBrowser_D/source/main.d
Error: unrecognized switch '-O3'
       run 'dmd -man' to open browser on manual
[6/8] Compiling D object 'gfontbrowser-test@exe/source_GFontBrowser.d.o'.
FAILED: gfontbrowser-test@exe/source_GFontBrowser.d.o 
/usr/bin/dmd -Igfontbrowser-test@exe -I. -I../../Repositories/Git/Masters/GFontBrowser_D -I../../Repositories/Git/Masters/GFontBrowser_D/source -I/usr/include/d/gtkd-3/ -I/home/users/russel/.dub/packages/unit-threaded-0.7.50/unit-threaded/source/ -color=on -wi -O -release -O3 -J/home/users/russel/Repositories/Git/Masters/GFontBrowser_D -unittest   -of'gfontbrowser-test@exe/source_GFontBrowser.d.o' -c ../../Repositories/Git/Masters/GFontBrowser_D/source/GFontBrowser.d
Error: unrecognized switch '-O3'
       run 'dmd -man' to open browser on manual
ninja: build stopped: subcommand failed.
```
Clearly this stems from the meson.build file which has:
```
project(
  'GFontBrowser',
  'd',
  version: '0.0.0',
  default_options: ['buildtype=release'],
)
```
but it appears it is LDC release options that are used even when DC is not LDC.

I retried the "force using dmd throughout" build:
```
|> DC=/usr/bin/dmd meson_head --prefix=$HOME/Built ~/Repositories/Git/Masters/GFontBrowser_D/ 
The Meson build system
Version: 0.48.0.dev1
Source dir: /home/users/russel/Repositories/Git/Masters/GFontBrowser_D
Build dir: /home/users/russel/BuildArea/GFontBrowser_D
Build type: native build
Project name: GFontBrowser
Project version: 0.0.0
Native D compiler: /usr/bin/dmd (dmd 2.081.1 "DMD64 D Compiler v2.081.1")
Build machine cpu family: x86_64
Build machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (0.29)
Dependency gtkd-3 found: YES 3.8.3
Found DUB: /usr/bin/dub (DUB version 1.10.0, built on Jul  3 2018)
Dependency unit-threaded found: YES 0.7.51
Build targets in project: 2
Found ninja-1.8.2 at /usr/bin/ninja
 
|> ninja
[0/1] Regenerating build files.
The Meson build system
Version: 0.48.0.dev1
Source dir: /home/users/russel/Repositories/Git/Masters/GFontBrowser_D
Build dir: /home/users/russel/BuildArea/GFontBrowser_D
Build type: native build
Project name: GFontBrowser
Project version: 0.0.0
Native D compiler: /usr/bin/dmd (dmd 2.081.1 "DMD64 D Compiler v2.081.1")
Build machine cpu family: x86_64
Build machine cpu: x86_64
Dependency gtkd-3 found: YES (cached)
Dependency unit-threaded found: YES (cached)
Build targets in project: 2
Found ninja-1.8.2 at /usr/bin/ninja
[7/8] Linking target gfontbrowser-test.
FAILED: gfontbrowser-test 
/usr/bin/dmd  -ofgfontbrowser-test 'gfontbrowser-test@exe/source_main.d.o' 'gfontbrowser-test@exe/source_About.d.o' 'gfontbrowser-test@exe/source_GFontBrowser.d.o' -main -L-L/usr/lib/x86_64-linux-gnu/ -L-lgtkd-3 -L-ldl /home/users/russel/.dub/packages/unit-threaded-0.7.51/unit-threaded/libunit-threaded.a  
/usr/bin/ld: /home/users/russel/.dub/packages/unit-threaded-0.7.51/unit-threaded/libunit-threaded.a(unit_threaded.should.o): undefined reference to symbol '_d_allocclass'
/usr/bin/ld: //usr/lib/x86_64-linux-gnu/libdruntime-ldc-shared.so.78: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
Error: linker exited with status 1
[8/8] Linking target gfontbrowser.
ninja: build stopped: subcommand failed.
```

--- a/ciimage/Dockerfile
+++ b/ciimage/Dockerfile
@@ -13,9 +13,13 @@ RUN apt-get -y update && apt-get -y upgrade \
 && apt-get -y install python-dev \
 && apt-get -y install libomp-dev openssh-client \
 && apt-get -y install -y clang libclang-dev llvm-dev flex \
+&& apt-get -y install gdc ldc \
 && python3 -m pip install hotdoc codecov \
 && dub fetch urld \
-&& dub build urld --compiler=gdc
+&& dub build urld --compiler=gdc \
+&& dub fetch dubtestproject \
+&& dub build dubtestproject:test1 --compiler=ldc2 \
+&& dub build dubtestproject:test2 --compiler=ldc2
 
 # OpenSSH client is needed to run openmpi binaries.
 

--- a/mesonbuild/compilers/d.py
+++ b/mesonbuild/compilers/d.py
@@ -95,6 +95,9 @@ class DCompiler(Compiler):
     def name_string(self):
         return ' '.join(self.exelist)
 
+    def get_exelist(self):
+        return self.exelist
+
     def get_linker_exelist(self):
         return self.exelist[:]
 
@@ -294,6 +297,11 @@ class DCompiler(Compiler):
                 # a linker search path.
                 dcargs.append('-L' + arg)
                 continue
+            elif arg.startswith('/') or arg.startswith('./'):
+                # absolute (or relative) paths passed to the linker may be static libraries
+                # or other objects that we need to link.
+                dcargs.append('-L' + arg)
+                continue
             elif arg.startswith('-mscrtlib='):
                 mscrtlib = arg[10:].lower()
 
@@ -309,6 +317,7 @@ class DCompiler(Compiler):
                         dcargs.append('-L/DEFAULTLIB:legacy_stdio_definitions.lib')
 
                 dcargs.append(arg)
+
                 continue
             dcargs.append(arg)
 

--- a/mesonbuild/dependencies/base.py
+++ b/mesonbuild/dependencies/base.py
@@ -24,6 +24,7 @@ import json
 import shlex
 import shutil
 import textwrap
+import platform
 from enum import Enum
 from pathlib import PurePath
 
@@ -849,12 +850,13 @@ class DubDependency(ExternalDependency):
         super().__init__('dub', environment, 'd', kwargs)
         self.name = name
         self.compiler = super().get_compiler()
+        self.module_path = None
 
         if 'required' in kwargs:
             self.required = kwargs.get('required')
 
         if DubDependency.class_dubbin is None:
-            self.dubbin = self.check_dub()
+            self.dubbin = self._check_dub()
             DubDependency.class_dubbin = self.dubbin
         else:
             self.dubbin = DubDependency.class_dubbin
@@ -875,20 +877,33 @@ class DubDependency(ExternalDependency):
             self.is_found = False
             return
 
-        j = json.loads(res)
         comp = self.compiler.get_id().replace('llvm', 'ldc').replace('gcc', 'gdc')
+        packages = []
+        j = json.loads(res)
         for package in j['packages']:
+            packages.append(package['name'])
             if package['name'] == name:
-                if j['compiler'] != comp:
-                    msg = ['Dependency', mlog.bold(name), 'found but it was compiled with']
-                    msg += [mlog.bold(j['compiler']), 'and we are using', mlog.bold(comp)]
-                    mlog.error(*msg)
+                self.is_found = True
+
+                not_lib = True
+                if 'targetType' in package:
+                    if package['targetType'] == 'library':
+                        not_lib = False
+
+                if not_lib:
+                    mlog.error(mlog.bold(name), 'found but it isn\'t a library')
+                    self.is_found = False
+                    return
+
+                self.module_path = self._find_right_lib_path(package['path'], comp, j, True, package['targetFileName'])
+
+                if not os.path.exists(self.module_path):
+                    mlog.error(mlog.bold(name), 'found but it wasn\'t compiled with', mlog.bold(comp))
                     self.is_found = False
                     return
 
                 self.version = package['version']
                 self.pkg = package
-                break
 
         if self.pkg['targetFileName'].endswith('.a'):
             self.static = True
@@ -899,39 +914,93 @@ class DubDependency(ExternalDependency):
         for path in self.pkg['importPaths']:
             self.compile_args.append('-I' + os.path.join(self.pkg['path'], path))
 
-        self.link_args = []
+        self.link_args = self.raw_link_args = []
         for flag in self.pkg['lflags']:
             self.link_args.append(flag)
 
-        search_paths = []
-        search_paths.append(os.path.join(self.pkg['path'], self.pkg['targetPath']))
-        found, res = self.__search_paths(search_paths, self.pkg['targetFileName'])
-        for file in res:
-            self.link_args.append(file)
+        self.link_args.append(os.path.join(self.module_path, self.pkg['targetFileName']))
 
-        self.is_found = found
+        # Handle dependencies
+        libs = []
+
+        def add_lib_args(field_name, target):
+            if field_name in target['buildSettings']:
+                if os.name is 'nt':
+                    for lib in target['buildSettings'][field_name]:
+                        if lib not in target['buildSettings'][field_name]:
+                            libs.append(lib)
+                            # TODO: Use CCompiler.find_library to find the path of the .lib files
+                else: # posix and everything else
+                    for lib in target['buildSettings'][field_name]:
+                        if lib not in libs:
+                            libs.append(lib)
+                            pkgdep = PkgConfigDependency(lib, environment, {'required': 'true', 'silent': 'true'})
+                            for arg in pkgdep.get_compile_args():
+                                self.compile_args.append(arg)
+                            for arg in pkgdep.get_link_args():
+                                self.link_args.append(arg)
+                            for arg in pkgdep.get_link_args(raw=True):
+                                self.raw_link_args.append(arg)
+
+        for target in j['targets']:
+            if target['rootPackage'] in packages:
+                add_lib_args('libs', target)
+                add_lib_args('libs-{}'.format(platform.machine()), target)
+                for file in target['buildSettings']['linkerFiles']:
+                    self.link_args.append(self._find_right_lib_path(file, comp, j))
 
     def get_compiler(self):
         return self.compiler
 
-    def __search_paths(self, search_paths, target_file):
-        found = False
-        res = []
-        if target_file == '':
-            return True, res
-        for path in search_paths:
-                if os.path.isdir(path):
-                    for file in os.listdir(path):
-                        if file == target_file:
-                            res.append(os.path.join(path, file))
-                            found = True
-        return found, res
+    def _find_right_lib_path(self, default_path, comp, j, folder_only=False, file_name=''):
+        path = ''
+
+        module_build_path = lib_file_name = ''
+        if folder_only:
+            module_build_path = default_path
+            lib_file_name = file_name
+        else:
+            module_build_path = os.path.dirname(default_path)
+            lib_file_name = os.path.basename(default_path)
+
+        # Get D version implemented in the compiler
+        # gdc doesn't support this
+        ret, res = self._call_dubbin(['--version'])
+
+        if ret != 0:
+            mlog.error('Failed to run {!r}', mlog.bold(comp))
+            self.is_found = False
+            return
+
+        d_ver = re.search('v[0-9].[0-9][0-9][0-9].[0-9]', res) # Ex.: v2.081.2
+        if d_ver is not None:
+            d_ver = d_ver.group().rsplit('.', 1)[0].replace('v', '').replace('.', '') # Fix structure. Ex.: 2081
+        else:
+            d_ver = '' # gdc
+
+        # Ex.: library-debug-linux.posix-x86_64-ldc_2081-EF934983A3319F8F8FF2F0E107A363BA
+        build_name = 'library-{}-{}.{}-{}-{}_{}'.format(j['buildType'], j['platform'][0], j['platform'][1], platform.machine(), comp, d_ver)
+        for entry in os.listdir(os.path.join(module_build_path, '.dub', 'build')):
+            if entry.startswith(build_name):
+                for file in os.listdir(os.path.join(module_build_path, '.dub', 'build', entry)):
+                    if file == lib_file_name:
+                        if folder_only:
+                            path = os.path.join(module_build_path, '.dub', 'build', entry)
+                        else:
+                            path = os.path.join(module_build_path, '.dub', 'build', entry, lib_file_name)
+                        break
+
+        return path
 
     def _call_dubbin(self, args, env=None):
         p, out = Popen_safe(self.dubbin.get_command() + args, env=env)[0:2]
         return p.returncode, out.strip()
 
-    def check_dub(self):
+    def _call_copmbin(self, args, env=None):
+        p, out = Popen_safe(self.compiler.get_exelist() + args, env=env)[0:2]
+        return p.returncode, out.strip()
+
+    def _check_dub(self):
         dubbin = ExternalProgram('dub', silent=True)
         if dubbin.found():
             try:

--- a/test cases/failing/85 dub libray/meson.build
+++ b/test cases/failing/85 dub libray/meson.build
@@ -1,0 +1,3 @@
+project('dub', 'd', meson_version: '0.48.0')
+
+dependency('dubtestproject', method: 'dub') # Not library (none)

--- a/test cases/failing/86 dub executable/meson.build
+++ b/test cases/failing/86 dub executable/meson.build
@@ -1,0 +1,3 @@
+project('dub', 'd', meson_version: '0.48.0')
+
+dependency('dubtestproject:test1', method: 'dub') # Not library (executable)

--- a/test cases/failing/87 dub compiler/meson.build
+++ b/test cases/failing/87 dub compiler/meson.build
@@ -1,0 +1,3 @@
+project('dub', 'd', meson_version: '0.48.0')
+
+dependency('dubtestproject:test2', method: 'dub') # Compiler mismatch

